File: router/tree/test_tree.cc
/*
* Test for C++ RadixTree performance.
* `get` run 10000000 times, about 70 seconds
*/

#include <iostream>
#include <time.h>
#include <memory>

#include "tree.h"

using std::cout;
using std::endl;
using std::shared_ptr;
using std::make_shared;

using radixtree::RadixTree;
using radixtree::Request;
using radixtree::HandleFunc;

auto handle = [](shared_ptr<Request> req) {
  cout << "Handled! The params are:" << endl;
  for (auto &param : req->params)
    cout << param.first << ": " << param.second << endl;
};

RadixTree tree;
auto req = make_shared<Request>("/user/Lime/male/25", "GET");

void testWork() {
  bool pathExisted;
  HandleFunc handler;

  cout << tree.insert("/user", handle, {"GET"}) << endl;
  cout << tree.insert("/user/:name", handle, {"GET"}) << endl;
  cout << tree.insert("/user/:name", handle, {"POST"}) << endl;
  cout << tree.insert("/user/:name/:sex/:age", handle, {"GET"}) << endl;
  cout << tree.insert("/user/lime", handle, {"GET"}) << endl;
  cout << tree.insert("/src/*filename", handle, {"GET"}) << endl;
  cout << tree.insert("/src/image.png", handle, {"GET"}) << endl << endl;

  tie(pathExisted, handler, req->params) = tree.get(req->path, req->method);

  cout << pathExisted << endl;
  if (handler)
    handler(req);
}

void testPerformance() {
  auto start = clock();
  for (auto i = 0; i < 10000000; i++)
    tree.get(req->path, req->method);

  cout << "Total cost " << (clock() - start) * 1.0 / CLOCKS_PER_SEC
       << "seconds" << endl;
}

int main() {
  testWork();
  testPerformance();
}

File: router/tree/test_tree.py
# -*- coding: utf-8 -*-
"""
    Test script for Python RadixTree preformance.

    `get` run 10000000 times, CPython about 50 seconds, pypy about 3.7s
"""

import sys
import os
from timeit import timeit

sys.path.append(os.path.join(os.path.dirname(os.path.dirname(
                os.path.dirname(os.path.abspath(__file__))))))

from router.tree import RadixTree


class Request(object):
    def __init__(self, path, method):
        self.path = path
        self.method = method
        self.params = {}

    def __repr__(self):
        return '<Request path: {}, method: {}, params: {}>'.format(
            self.path, self.method, self.params)


def handle(req):
    print('Handled! The params are:')
    print(req.params)


def testWork():
    try:
        tree.insert("/user", handle, "GET")
        tree.insert("/user/:name", handle, "GET")
        tree.insert("/user/:name", handle, "POST")
        tree.insert("/user/:name/:sex/:age", handle, "GET")
        tree.insert("/user/lime", handle, "GET")
        tree.insert("/src/*filename", handle, "GET")
        tree.insert("/src/image.png", handle, "GET")
    except Exception as e:
        print(e)

    pathFound, handler, req.params = tree.get(req.path, req.method)
    if handler:
        handler(req)


def testPerformance():
    def f():
        for i in range(10000000):
            tree.get("/user/Lime/male/25", "GET")

    print(timeit(f, number=1))


tree = RadixTree()
req = Request("/user/Lime/male/25", "GET")


testWork()
testPerformance()
